### PR TITLE
core: lpae: mask exceptions during core_mmu_find_table()

### PR DESCRIPTION
**Pull request by jforissier**
core_mmu_find_table() calls get_core_pos() so it should mask interrupts
to avoid being re-scheduled to another core.

Signed-off-by: Jerome Forissier <jerome.forissier@linaro.org>

### CODE REVIEW

**Comment by etienne-lms** [comment]
This change looks fine.

A weird par is that we call this get mmu table, so we get the mmu table for a core, but then we unmask exception before accessing the mmu table, so we can have migrated to another core and when modifying the table. Sound strange: why bothering about the core used to get the table?
I think this work because all cores do share the very same tables, right?

**Comment by jforissier** [comment]
@etienne-lms I'm glad you asked because I though exactly the same thing as you!
If your explanation is correct, then I suppose we could replace get_core_pos() with 0?
Let's call @jenswi-linaro for help here :wink:

**Comment by jenswi-linaro** [comment]
No, `0` can't be used. If we're looking for a virtual address in user mode address range we'll traverse thread specific translation tables.

**Comment by jforissier** [comment]
@jenswi-linaro OK, in this case, shouldn't exceptions be masked by the caller of `core_mmu_find_table() ` instead?

**Comment by jenswi-linaro** [comment]
I'd rather not put that requirement on the caller. Perhaps it's enough to do something like this:
```
--- a/core/arch/arm/mm/core_mmu_lpae.c
+++ b/core/arch/arm/mm/core_mmu_lpae.c
@@ -666,6 +666,17 @@ bool core_mmu_find_table(vaddr_t va, unsigned max_level,
 			 * We've either reached max_level, level 3, a block
 			 * mapping entry or an "invalid" mapping entry.
 			 */
+
+			/*
+			 * Level 1 is the CPU specific translation table.
+			 * It doesn't make sense to return anything based
+			 * on that unless foreign interrupts already are
+			 * masked.
+			 */
+			if (level == 1 &&
+			    !(exceptions & THREAD_EXCP_FOREIGN_INTR))
+				goto out
+
 			tbl_info->table = tbl;
 			tbl_info->va_base = va_base;
 			tbl_info->level = level;
```

**Comment by etienne-lms** [comment]
Looks nice proposal.
Maybe add:
```c
			if (level == 1 &&
			    !(exceptions & THREAD_EXCP_FOREIGN_INTR)) {
+				assert(max_level == 1);
				goto out
			}

```

**Comment by jenswi-linaro** [comment]
Why?

**Comment by etienne-lms** [comment]
i'm dumb. call will nicely fail if it are not masked. that's fine.
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/mm/core_mmu_lpae.c [hunk 1]
```diff
@@ -642,46 +642,64 @@ void core_mmu_create_user_map(struct user_ta_ctx *utc,
...
 			 */
+
+			/*
+			 * Level 1 is the CPU specific translation table.
+			 * It doesn't make sense to return anything based
+			 * on that unless foreign interrupts already are
+			 * masked.
+			 */
+			if (level == 1 &&
+			    !(exceptions & THREAD_EXCP_FOREIGN_INTR))
```
Comment (vchong) [inline core/arch/arm/mm/core_mmu_lpae.c]: Is `!(exceptions & THREAD_EXCP_FOREIGN_INTR)` really necessary? Everything is already masked when `exception` is initialized above.

Comment (etienne-lms) [reply core/arch/arm/mm/core_mmu_lpae.c]: `exceptions` gives the state of the exceptions masked before `core_mmu_find_table()` called `thread_mask_exceptions()`. This is what we wanted to test.

Comment (vchong) [reply core/arch/arm/mm/core_mmu_lpae.c]: Ah, ok. Thanks!

**Pull request merged**